Simplify recovery logic in IndicesClusterStateService

ywelsch opened the pull request:
- Moves recovery logic into IndexShard
- Lets recoveries be cancelled by shard closing
- Ensures routingEntry is set on initialization of IndexShard

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -274,6 +274,10 @@ public boolean allocatedPostIndexCreate(IndexMetaData indexMetaData) {
     }
 
+    public boolean isPeerRecovery() {
+        return !primary() || relocatingNodeId() != null;
```
channeling someone by saying `primary() == false` :)

bleskes replied:
can we change to use isRelocationTarget() instead of relocatingNodeId() != null ? I think it's confusing that a relocation source will return true here.

bleskes commented on core/src/main/java/org/elasticsearch/index/IndexService.java:
```diff
@@ -336,18 +336,17 @@ public synchronized IndexShard createShard(ShardRouting routing) throws IOExcept
             }
             eventListener.indexShardStateChanged(indexShard, null, indexShard.state(), "shard created");
             eventListener.afterIndexShardCreated(indexShard);
-            indexShard.updateRoutingEntry(routing, true);
```
💯

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1326,6 +1330,58 @@ protected Engine getEngineOrNull() {
         return this.currentEngineReference.get();
     }
 
+    public void startRecovery(ShardRouting shardRouting, DiscoveryNode localNode, DiscoveryNode sourceNode, RecoveryTargetService recoveryTargetService, RecoveryTargetService.RecoveryListener recoveryListener, RepositoriesService repositoriesService) {
```
nite: line length..

bleskes replied:
why do we need to pass the shardRouting here? can't we use the one of this IndexShard?

ywelsch replied:
yes, we can use the IndexShard one.

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1326,6 +1330,58 @@ protected Engine getEngineOrNull() {
+                recoveryListener.onRecoveryFailure(recoveryState, new RecoveryFailedException(shardId, sourceNode, localNode, e), true);
+            }
+        } else if (restoreSource == null) {
+            assert routingEntry().equals(shardRouting); // should have already be done before
```
see comment above

bleskes commented on core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java:
```diff
@@ -93,7 +92,7 @@
     private final NodeServicesProvider nodeServicesProvider;
 
-    private static final ShardStateAction.Listener SHARD_STATE_ACTION_LISTENER = new ShardStateAction.Listener() {
+    public static final ShardStateAction.Listener SHARD_STATE_ACTION_LISTENER = new ShardStateAction.Listener() {
```
It seems this can go back to private?

ywelsch replied:
yup

bleskes commented on core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java:
```diff
@@ -525,21 +524,22 @@ private void applyNewOrUpdatedShards(final ClusterChangedEvent event) {
+                if (shardRouting.isPeerRecovery()) {
+                    RecoveryState recoveryState = indexShard.recoveryState();
+                    if (recoveryState != null && recoveryState.getStage() != RecoveryState.Stage.DONE) { // shard is still recovering
+                        if (recoveryState.getType() == RecoveryState.Type.PRIMARY_RELOCATION || recoveryState.getType() == RecoveryState.Type.REPLICA) {
```
isn't this what peer recovery means? should we just check currentRoutingEntry?  at this point we know it has to be the same allocationId, which will means it's safe.

ywelsch replied:
The condition came to be like that as I saw a test failure where primary shards that were a relocation source were being failed. The check ensured that this shard was NOT previously recovered from store. Why was this check needed? If we are a primary shard that is relocation source (i.e. recovered from store), then we cancel the shard as the relocation target is wrongly taken as source node whereas the current node is taking as previous source. The reason for this is that `ShardRouting.isPeerRecovery` returns true when we have a primary that is a relocation source. I've now done what I should have done in the first place, adapting `ShardRouting.isPeerRecovery` to ensure that it only applies to primary that is relocation target.

bleskes commented on core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java:
```diff
@@ -525,21 +524,22 @@ private void applyNewOrUpdatedShards(final ClusterChangedEvent event) {
+                    if (recoveryState != null && recoveryState.getStage() != RecoveryState.Stage.DONE) { // shard is still recovering
+                        if (recoveryState.getType() == RecoveryState.Type.PRIMARY_RELOCATION || recoveryState.getType() == RecoveryState.Type.REPLICA) {
+                            final DiscoveryNode sourceNode = findSourceNodeForPeerRecovery(logger, routingTable, nodes, shardRouting);
+                            if (recoveryState.getSourceNode().equals(sourceNode) == false) {
```
we lose some concurrency control  here where we only cancel the recovery if it's not done. I'm wondering if we should use `cancelRecoveriesForShard` and if it returns  true, we then remove the shard. 

Also - In all cases where the the source of recovery changes due to a primary failure, the master cancels the allocation and changes the allocation id, meaning we don't get here. I wonder if we should also catch the case where a primary relocates on the master and not have to worry about all of this here. I think it will be simpler all in all

ywelsch replied:
good idea. I've taken the cancelRecoveriesForShard approach.

>  I wonder if we should also catch the case where a primary relocates on the master and not have to worry about all of this here. I think it will be simpler all in all.

I'm less convinced of that one. Fully handling it on the master means that a shard that is in POST_RECOVERY can be cancelled just because primary relocation completed before shard was activated by the master node. This is especially bad if there is some load on master (with longer queue of tasks) where shards take a bit of time to get started and relocation of primary is going on. I think that it is sufficient for master to determine the recovery source for each shard. We can then easily check if it needs to be changed.

bleskes replied:
> can be cancelled just because primary relocation completed before shard was activated by the master node

yes. I'm aware of that - I'm thinking that with seq# fast recovery it wouldn't matter much as a ready shard will quickly re-recover. However, seeing how the new code looks like with the cancelRecoveriesForShard + shardRouting.isPeerRecovery changes, I think it became much simpler. I'm good. We can see how things develop later on and potentially move some logic to the master (which will simplify this class) or not.

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java:
```diff
@@ -136,36 +136,21 @@ public int size() {
             if (status.shardId().equals(shardId)) {
-                boolean cancel = false;
                 // if we can't increment the status, the recovery is not there any more.
                 if (status.tryIncRef()) {
```
I think this tryIncRef is only  there to make sure shouldCancel can called without things changing on it. I think we can just call cancelRecovery and see whether it returns true.

bleskes commented:
I like the change. Left some comments/ suggestions

ywelsch commented:
@bleskes I've reworked the bit where recovery source changed. Please have another look. I've also added a commit that I forgot to push as part of this PR that fixes an issue in IndexShardTests when `UnassignedInfo.Reason` is `INDEX_CREATED` (no recovery happens for shards with `ShardRouting.allocatedPostIndexCreate == false`).

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -275,7 +275,7 @@ public boolean allocatedPostIndexCreate(IndexMetaData indexMetaData) {
 
     public boolean isPeerRecovery() {
-        return !primary() || relocatingNodeId() != null;
+        return state == ShardRoutingState.INITIALIZING && (primary() == false || relocatingNodeId != null);
```
can we change to use isRelocationTarget() instead?

ywelsch replied:
I prefer it this way as I see the `state == ShardRoutingState.INITIALIZING` condition as essential to both parts of the condition (it should also hold for non-primary shards). Making it 
`(state == ShardRoutingState.INITIALIZING && primary() == false) || isRelocationTarget()` does not read better...

bleskes replied:
with the change to only consider initializing shards, I agree - it's more explicit and better. can we add java docs to the method that it only returns true for intializing shards  that should pull their data from a remote copy?

ywelsch replied:
will do

bleskes commented on core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java:
```diff
@@ -526,15 +526,13 @@ private void applyNewOrUpdatedShards(final ClusterChangedEvent event) {
-                            }
+                    final DiscoveryNode sourceNode = findSourceNodeForPeerRecovery(logger, routingTable, nodes, shardRouting);
+                    if (recoveryState.getSourceNode().equals(sourceNode) == false) {
+                        if (recoveryTargetService.cancelRecoveriesForShard(currentRoutingEntry.shardId(), "recovery source node changed")) {
```
+1

bleskes commented:
LGTM. Left one minor request for docs.

ywelsch commented:
Thanks for the review @bleskes!
